crmsh/migration.py

```diff
@@ -0,0 +1,377 @@
+        raise NotImplementedError
+
+    def handle_problem(self, is_fatal: bool, title: str, detail: typing.Iterable[str]):
+        raise NotImplementedError
+
+    def end(self):
+        raise NotImplementedError
+
+
+class CheckResultJsonHandler(CheckResultHandler):
+    def __init__(self, indent: typing.Optional[int] = None):
+        self._indent = indent
+        self.json_result = {
+            "pass": True,
+            "problems": [],
+            "tips": [],
+        }
+    def log_info(self, fmt: str, *args):
+        logger.debug(fmt, *args)
+
+    def handle_tip(self, title: str, details: typing.Iterable[str]):
+        self.json_result["tips"].append({
+            "title": title,
+            "descriptions": details if isinstance(details, list) else list(details),
+        })
+
+    def handle_problem(self, is_fatal: bool, title: str, detail: typing.Iterable[str]):
+        self.json_result["pass"] = False
+        self.json_result["problems"].append({
+            "is_fatal": is_fatal,
+            "title": title,
+            "descriptions": detail if isinstance(detail, list) else list(detail),
+        })
+
+    def end(self):
+        json.dump(
+            self.json_result,
+            sys.stdout,
+            ensure_ascii=False,
+            indent=self._indent,
+        )
+        sys.stdout.write('\n')
+
+
+class CheckResultInteractiveHandler(CheckResultHandler):
+    def __init__(self):
+        self.has_problems = False
+
+    def log_info(self, fmt: str, *args):
+        self.write_in_color(sys.stdout, constants.GREEN, '[INFO] ')
+        print(fmt % args)
+
+    def handle_problem(self, is_fatal: bool, title: str, details: typing.Iterable[str]):
+        self.has_problems = True
+        self.write_in_color(sys.stdout, constants.YELLOW, '[FAIL] ')
+        print(title)
+        for line in details:
+            sys.stdout.write('       ')
+            print(line)
+        if is_fatal:
+            raise MigrationFailure('Unable to start migration.')
+
+    def handle_tip(self, title: str, details: typing.Iterable[str]):
+        self.write_in_color(sys.stdout, constants.YELLOW, '[WARN] ')
+        print(title)
+        for line in details:
+            sys.stdout.write('       ')
+            print(line)
+
+    @staticmethod
+    def write_in_color(f, color: str, text: str):
+        if f.isatty():
+            f.write(color)
+            f.write(text)
+            f.write(constants.END)
+        else:
+            f.write(text)
+
+    def end(self):
+        if self.has_problems:
+            self.write_in_color(sys.stdout, constants.RED, '[FAIL]')
+            sys.stdout.write(' Please fix all the "FAIL" problems above before migrating to SLES 16.\n\n')
+        else:
+            self.write_in_color(sys.stdout, constants.GREEN, '[PASS]')
+            sys.stdout.write(' Good to migrate to SLES 16.\n\n')
+
+
+def check(args: typing.Sequence[str]) -> int:
```

Comment:
It is better to add a docstring to:
- explain what the `--json` and `--local` option are used for
- introduce this function and return code

Reply:
This is quite self-explained. Adding docstrings like "migration.check: run migration checks", or "--json    output the result in json format" does not make sense. And the return code is also the common main() style return code.